src/sql/base/browser/ui/taskbar/overflowActionbar.ts

```diff
@@ -0,0 +1,297 @@
+	private _moreActionsElement: HTMLElement;
+
+	constructor(container: HTMLElement, options: IActionBarOptions = defaultOptions) {
+		super(container, options);
+
+		this._register(DOM.addDisposableListener(window, DOM.EventType.RESIZE, e => {
+			if (this._actionsList) {
+				this.resizeToolbar();
+			}
+		}));
+
+		this._overflow = document.createElement('ul');
+		this._overflow.id = 'overflow';
+		this._overflow.className = 'overflow';
+		this._overflow.setAttribute('role', 'menu');
+		this._domNode.appendChild(this._overflow);
+
+		this._actionsList.style.flexWrap = 'nowrap';
+
+		container.appendChild(this._domNode);
+	}
+
+	@debounce(300)
+	private resizeToolbar() {
+		let width = this._actionsList.offsetWidth;
+		let fullWidth = this._actionsList.scrollWidth;
+
+		// collapse actions that are beyond the width of the toolbar
+		if (width < fullWidth) {
+			// create '•••' more element if it doesn't exist yet
+			if (!this._moreItemElement) {
+				this.createMoreItemElement();
+			}
+
+			this._moreItemElement.style.display = 'block';
+			while (width < fullWidth) {
+				let index = this._actionsList.childNodes.length - 2; // remove the last toolbar action before the more actions '...'
+				if (index > -1) {
+					// move placeholder in this._items
+					this.collapseItem();
+					fullWidth = this._actionsList.scrollWidth;
+				} else {
+					break;
+				}
+			}
+		} else if (this._overflow?.hasChildNodes()) { // uncollapse actions if there is space for it
+			while (width === fullWidth && this._overflow.hasChildNodes()) {
+				// move placeholder in this._items
+				let placeHolderItem = this._items.splice(this._actionsList.childNodes.length - 1, 1);
+				this._items.splice(this._actionsList.childNodes.length, 0, placeHolderItem[0]);
+
+				let item = this._overflow.removeChild(this._overflow.firstChild);
+				// change role back to button when it's in the toolbar
+				if ((<HTMLElement>item).className !== 'taskbarSeparator') {
+					(<HTMLElement>item.firstChild).setAttribute('role', 'button');
+				}
+				this._actionsList.insertBefore(item, this._actionsList.lastChild);
+
+				// if the action was too wide, collapse it again
+				if (this._actionsList.scrollWidth > this._actionsList.offsetWidth) {
+					// move placeholder in this._items
+					this.collapseItem();
+					break;
+				} else if (!this._overflow.hasChildNodes()) {
+					this._moreItemElement.style.display = 'none';
+				}
+			}
+		}
+	}
+
+	private collapseItem(): void {
+		// move placeholder in this._items
+		let placeHolderItem = this._items.splice(this._actionsList.childNodes.length - 1, 1);
+		this._items.splice(this._actionsList.childNodes.length - 2, 0, placeHolderItem[0]);
+
+		let index = this._actionsList.childNodes.length - 2; // remove the last toolbar action before the more actions '...'
+		let item = this._actionsList.removeChild(this._actionsList.childNodes[index]);
+		this._overflow.insertBefore(item, this._overflow.firstChild);
```

Comment:
how do we ensure the group division is maintained? like between restore and refresh in the gif?

Reply:
Separators are also added to the overflow and become horizontal, like this: ![image](https://user-images.githubusercontent.com/31145923/78848087-909dcb80-79c5-11ea-937d-1fe0d87b5907.png)